md-proto/WebAssembly.md

```diff
@@ -0,0 +1,2007 @@
+[validated](#validation).
+
+The contents of the [Data Section](#data-section) are loaded into linear memory.
+Each [string] is loaded into linear memory at its associated offset.
+
+Then, the [recursion limit](#recursion-limits) values are chosen,
+[nondeterministically].
+
+TODO: Describe module versus instance, and describe instantiation.
+
+TODO: Implementations can, nondeterministically, trap at any time, due to
+insufficient resources.
+
+### Recursion Limits
+
+The following two values are chosen before any functions are executed:
+
+ - A *recursion increment*.
+ - A *recursion limit*.
+
+> These values are used by [call instructions][L].
+
+### Module Execution
+
+If the module contains a [Start Section](#start-section), the referenced
+function is [executed](#function-execution).
+
+### Function Execution
+
+Function execution can be prompted by a [call-family instruction][L], by
+[module execution](#module-execution), or by the embedding environment.
+
+The input to execution of a function consists of:
+ - The function to be executed.
+ - The incoming argument values, one for each argument [type] of the function.
+ - A *recursion depth* value.
+
+For the duration of the execution of a function body, several data structures
+are created:
+ - A *control-flow stack*, which holds [labels] for reference from branch
+   instructions.
+ - A *value stack*, which carries values between instructions.
+ - A *locals* array, containing an element for each argument of the function in
+   the argument's [type], as well as an element for each local declaration in
+   the function.
+ - A *current position*.
+
+> Implementations need not create a literal array to store the locals, or
+literal stacks to manage values at runtime.
+
+#### Function Execution Initialization
+
+The current position starts at the first instruction in the function body. The
+value stack begins empty. The control-flow stack begins with an entry holding a
+[label] bound to the last instruction in the instruction sequence.
+
+The value of each incoming argument is copied to the local with the
+corresponding index, and the rest of the locals are initialized to all-zeros
+bit-pattern values.
+
+#### Function Body Execution
+
+The instruction at the current position is remembered, and the current position
+is incremented to point to the position following it. Then the remembered
+instruction is executed as follows.
+
+For each operand [type] in the instruction's signature in reverse order, a value
+is popped from the value stack and provided as the corresponding operand value.
+The instruction is then executed as described in the
+[Instructions Section](#instructions) entry describing it. Each of the
+instruction's return values are then pushed onto the value stack.
+
+If the current position is now past the end of the sequence,
+[function return execution](#function-return-execution) is initiated and
+execution of the function is thereafter complete.
+
```

Comment:
Why not check that values are not popped past the control-flow blocks top-of-stack? It seems really odd that this would allow crossing a control-flow stack-top boundary when popping arguments, and then allow the results to be pushed back, and only then check if the final stack position against the control-flow stack-top boundary?